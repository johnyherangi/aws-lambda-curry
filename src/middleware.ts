import { Context } from "aws-lambda"
import { Handler } from "./handler"

export type Middleware<TEvent, TResult> = (
    event: TEvent,
    context: Context,
    next: Handler<TEvent, TResult>,
) => TResult | Promise<TResult>

export function middleware<TEvent, TResult>(middlewares?: Middleware<TEvent, TResult>[]) {
    const invoke = (
        event: TEvent,
        context: Context,
        handler: Handler<TEvent, TResult>,
        pipeline?: Middleware<TEvent, TResult>[],
    ): TResult | Promise<TResult> => {
        if (!pipeline) {
            return handler(event, context)
        }

        const pipelineCopy = [...pipeline]
        const middleware = pipelineCopy.pop()

        if (!middleware) {
            return handler(event, context)
        }

        const next: Handler<TEvent, TResult> = (e, c) => {
            return invoke(e, c, handler, pipelineCopy)
        }

        return middleware(event, context, next)
    }

    return (handler: Handler<TEvent, TResult>): Handler<TEvent, TResult> => {
        return (e, c) => invoke(e, c, handler, middlewares)
    }
}
