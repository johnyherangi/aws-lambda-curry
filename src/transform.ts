import { Context } from "aws-lambda"
import { Handler } from "./handler"

export type Transform<TEvent, TResult> = (event: TEvent, context: Context) => TResult

export type TransformOutputs<TEvent, TResult, TTransforms extends Transform<TEvent, TResult>[]> = {
    [K in keyof TTransforms]: ReturnType<TTransforms[K]>
}

export type TransformHandler<TEvent, TResult, TTransform extends Transform<TEvent, TResult>[]> = (
    args: TransformOutputs<TEvent, TResult, TTransform>,
) => TResult | Promise<TResult>

export type TransformBuilder<TEvent, TResult, TTransforms extends Transform<TEvent, TResult>[]> = (
    transforms: TTransforms,
) => (handler: TransformHandler<TEvent, TResult, TTransforms>) => Handler<TEvent, TResult>

export type EmptyTransformBuilder<TEvent, TResult> = () => (
    handler: Handler<TEvent, TResult>,
) => Handler<TEvent, TResult>

export function transform<TEvent, TResult, TTransforms extends Transform<TEvent, any>[]>(
    transforms: TTransforms,
): (handler: TransformHandler<TEvent, TResult, TTransforms>) => Handler<TEvent, TResult>

export function transform<TEvent, TResult>(): (
    handler: Handler<TEvent, TResult>,
) => Handler<TEvent, TResult>

export function transform<TTransforms extends Transform<TEvent, any>[], TEvent, TResult>(
    transforms?: TTransforms,
) {
    if (!transforms) {
        return function (handler: Handler<TEvent, TResult>) {
            return handler
        }
    }

    return function (handler: TransformHandler<TEvent, TResult, TTransforms>) {
        return function (event: TEvent, context: Context) {
            const args = transforms?.map((t) => t(event, context)) as any
            return handler(args)
        }
    }
}
