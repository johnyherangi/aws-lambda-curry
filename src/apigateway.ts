import { APIGatewayProxyEvent, APIGatewayProxyResult } from "aws-lambda"
import { Handler } from "./handler"
import { Middleware, middleware } from "./middleware"
import { Transform, TransformHandler, transform } from "./transform"

export function apigateway(
    middlewares?: Middleware<APIGatewayProxyEvent, APIGatewayProxyResult>[],
): <TTransforms extends Transform<APIGatewayProxyEvent, any>[]>(
    transforms: TTransforms,
) => (
    handler: TransformHandler<APIGatewayProxyEvent, APIGatewayProxyResult, TTransforms>,
) => Handler<APIGatewayProxyEvent, APIGatewayProxyResult>

export function apigateway(
    middlewares?: Middleware<APIGatewayProxyEvent, APIGatewayProxyResult>[],
): () => (
    handler: Handler<APIGatewayProxyEvent, APIGatewayProxyResult>,
) => Handler<APIGatewayProxyEvent, APIGatewayProxyResult>

export function apigateway(
    middlewares?: Middleware<APIGatewayProxyEvent, APIGatewayProxyResult>[],
) {
    function transformFn<TTransforms extends Transform<APIGatewayProxyEvent, any>[]>(
        transforms: TTransforms,
    ): (
        handler: TransformHandler<APIGatewayProxyEvent, APIGatewayProxyResult, TTransforms>,
    ) => Handler<APIGatewayProxyEvent, APIGatewayProxyResult>

    function transformFn(): (
        handler: Handler<APIGatewayProxyEvent, APIGatewayProxyResult>,
    ) => Handler<APIGatewayProxyEvent, APIGatewayProxyResult>

    function transformFn<TTransforms extends Transform<APIGatewayProxyEvent, any>[]>(
        transforms?: TTransforms,
    ) {
        if (!transforms) {
            return function (handler: Handler<APIGatewayProxyEvent, APIGatewayProxyResult>) {
                return middleware(middlewares)(handler)
            }
        }

        return function (
            handler: TransformHandler<APIGatewayProxyEvent, APIGatewayProxyResult, TTransforms>,
        ) {
            return middleware(middlewares)(
                transform<APIGatewayProxyEvent, APIGatewayProxyResult, TTransforms>(
                    transforms ?? ([] as any),
                )(handler),
            )
        }
    }

    return transformFn
}
