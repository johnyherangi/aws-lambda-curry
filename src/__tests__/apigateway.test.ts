import { apigateway } from "@src/apigateway"
import { Middleware } from "@src/middleware"
import { Transform } from "@src/transform"
import { APIGatewayProxyEvent, APIGatewayProxyResult, Context } from "aws-lambda"

describe("apigateway.ts", () => {
    describe("apigateway()", () => {
        it("creates apigateway handler", () => {
            const getPath: Transform<APIGatewayProxyEvent, string> = (e, c) => e.path

            const middleware1: Middleware<APIGatewayProxyEvent, APIGatewayProxyResult> = (
                e,
                c,
                next,
            ) => next(e, c)

            const handler = apigateway([middleware1])([getPath])(([path]) => {
                return {
                    statusCode: 200,
                    body: path,
                }
            })

            const event = {
                httpMethod: "GET",
                path: "/example",
            } as APIGatewayProxyEvent

            expect(handler(event, {} as Context)).toMatchSnapshot()
        })
        it("accepts no transforms", () => {
            const handler = apigateway()()((event) => {
                return {
                    statusCode: 200,
                    body: event.path,
                }
            })

            const event = {
                httpMethod: "GET",
                path: "/example",
            } as APIGatewayProxyEvent

            expect(handler(event, {} as Context)).toMatchSnapshot()
        })
    })
})
