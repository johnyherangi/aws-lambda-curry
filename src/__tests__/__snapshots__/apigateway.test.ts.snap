// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apigateway.ts apigateway() creates apigateway handler 1`] = `
{
  "body": "/example",
  "statusCode": 200,
}
`;
